pkg/controller/common/association/conf.go

```diff
@@ -0,0 +1,135 @@
+	return true, nil
+}
+
+// GetAssociationConf extracts the association configuration from the given object by reading the annotations.
+func GetAssociationConf(obj runtime.Object) (*commonv1alpha1.AssociationConf, error) {
+	accessor := meta.NewAccessor()
+	annotations, err := accessor.Annotations(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	return extractAssociationConf(annotations)
+}
+
+func extractAssociationConf(annotations map[string]string) (*commonv1alpha1.AssociationConf, error) {
+	if len(annotations) == 0 {
+		return nil, nil
+	}
+
+	var assocConf commonv1alpha1.AssociationConf
+	serializedConf, exists := annotations[annotation.AssociationConfAnnotation]
+	if !exists || serializedConf == "" {
+		return nil, nil
+	}
+
+	if err := json.Unmarshal(unsafeStringToBytes(serializedConf), &assocConf); err != nil {
+		return nil, errors.Wrapf(err, "failed to extract association configuration")
+	}
+
+	return &assocConf, nil
+}
+
+// RemoveAssociationConf removes the association configuration annotation.
+func RemoveAssociationConf(client k8s.Client, obj runtime.Object) error {
+	accessor := meta.NewAccessor()
+	annotations, err := accessor.Annotations(obj)
+	if err != nil {
+		return err
+	}
+
+	if len(annotations) == 0 {
+		return nil
+	}
+
+	if _, exists := annotations[annotation.AssociationConfAnnotation]; !exists {
+		return nil
+	}
+
+	delete(annotations, annotation.AssociationConfAnnotation)
+	if err := accessor.SetAnnotations(obj, annotations); err != nil {
+		return err
+	}
+
+	return client.Update(obj)
+}
+
+// UpdateAssociationConf updates the association configuration annotation.
+func UpdateAssociationConf(client k8s.Client, obj runtime.Object, wantConf *commonv1alpha1.AssociationConf) error {
+	accessor := meta.NewAccessor()
+	annotations, err := accessor.Annotations(obj)
+	if err != nil {
+		return err
+	}
+
+	// serialize the config and update the object
+	serializedConf, err := json.Marshal(wantConf)
+	if err != nil {
+		return errors.Wrapf(err, "failed to serialize configuration")
+	}
+
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	annotations[annotation.AssociationConfAnnotation] = unsafeBytesToString(serializedConf)
+	if err := accessor.SetAnnotations(obj, annotations); err != nil {
+		return err
+	}
+
+	// persist the changes
+	return client.Update(obj)
+}
+
+// unsafeStringToBytes converts a string to a byte array without making extra allocations.
+// since we read potentially large strings from annotations on every reconcile loop, this should help
+// reduce the amount of garbage created.
+func unsafeStringToBytes(s string) []byte {
+	hdr := *(*reflect.StringHeader)(unsafe.Pointer(&s))
+	return *(*[]byte)(unsafe.Pointer(&reflect.SliceHeader{
+		Data: hdr.Data,
+		Len:  hdr.Len,
+		Cap:  hdr.Len,
+	}))
+}
+
+// unsafeBytesToString converts a byte array to string without making extra allocations.
+func unsafeBytesToString(b []byte) string {
+	return *(*string)(unsafe.Pointer(&b))
+}
```

Comment:
Probably not. It's just force of habit to me. I don't think there's a downside to keeping them but happy to reconsider if you feel otherwise.

Reply:
I think I'm fine using this kind of optimization, my only concern is to have it in the `association` package, maybe `stringsutil` would be more appropriate ?

Reply:
I am not sure whether the unsafe functions deserve to be publicly visible utility functions in `stringsutil` because they require careful consideration before use. My inclination is to keep them localised here for the time being and if we find ourselves needing them elsewhere, only then consider moving them to a utils package. WDYT?

Reply:
I think I agree with you. Would be nice to add a note in the comment about it.